File: iosApp/iosApp/Llm/LlmBridge.swift
import Foundation
import Shared
import LiteRTLM

/// LiteRT-LM Swift runtime behind the shared `NativeLlmBridge` contract.
/// Shared Kotlin code drives it through callbacks; tool calls made by the
/// model are routed back into shared code via `MainViewControllerKt.executeBuiltinTool`.
final class SwiftLlmBridge: NSObject, NativeLlmBridge {

    static let shared = SwiftLlmBridge()

    /// Set by shared code (ChatSessionManager via Koin); tools call back into shared code instead.
    var toolExecutor: (any Shared.ToolExecutor)?

    private var engine: Engine?
    // Qualified: the Shared (Kotlin) framework also exports a `Conversation` type.
    private var conversations: [String: LiteRTLM.Conversation] = [:]
    private var generations: [String: Task<Void, Never>] = [:]

    // MARK: - Engine

    func initializeEngine(setup: EngineSetup, callback: any EngineCallback) {
        Task {
            do {
                self.teardownEngine()
                let config = try EngineConfig(
                    modelPath: setup.modelPath,
                    backend: setup.backend == "gpu" ? .gpu : .cpu(),
                    maxNumTokens: Int(setup.maxTokens),
                    cacheDir: setup.cacheDir
                )
                let engine = Engine(engineConfig: config)
                try await engine.initialize()
                self.engine = engine
                callback.onEngineReady()
            } catch {
                callback.onEngineError(message: describe(error))
            }
        }
    }

    func closeEngine() {
        teardownEngine()
    }

    private func teardownEngine() {
        generations.values.forEach { $0.cancel() }
        generations.removeAll()
        conversations.removeAll()
        engine = nil
    }

    // MARK: - Conversations

    func createConversation(setup: ConversationSetup, callback: any ConversationCallback) {
        Task {
            do {
                guard let engine = self.engine else {
                    callback.onConversationError(message: "Engine is not initialized")
                    return
                }
                let sampler = try SamplerConfig(
                    topK: Int(setup.topK),
                    topP: Float(setup.topP),
                    temperature: Float(setup.temperature)
                )
                let systemMessage = (setup.systemPrompt?.isEmpty == false)
                    ? Message(setup.systemPrompt!, role: .system)
                    : nil
                let config = ConversationConfig(
                    systemMessage: systemMessage,
                    initialMessages: self.initialMessages(fromHistoryJson: setup.historyJson),
                    tools: self.tools(for: setup.toolIds),
                    samplerConfig: sampler
                )
                let conversation = try await engine.createConversation(with: config)
                let handle = UUID().uuidString
                self.conversations[handle] = conversation
                callback.onConversationReady(handle: handle)
            } catch {
                callback.onConversationError(message: describe(error))
            }
        }
    }

    func closeConversation(handle: String) {
        generations[handle]?.cancel()
        generations[handle] = nil
        if let conversation = conversations.removeValue(forKey: handle) {
            try? conversation.cancel()
        }
    }

    // MARK: - Generation

    func sendMessage(handle: String, text: String, listener: any GenerationListener) {
        guard let conversation = conversations[handle] else {
            listener.onGenerationError(message: "Conversation is not active")
            return
        }
        let task = Task {
            do {
                for try await chunk in conversation.sendMessageStream(Message(text)) {
                    if Task.isCancelled { break }
                    let piece = chunk.toString
                    if !piece.isEmpty {
                        listener.onChunk(text: piece)
                    }
                    if let thought = chunk.channels["thought"], !thought.isEmpty {
                        listener.onThought(text: thought)
                    }
                }
                listener.onGenerationDone()
            } catch {
                if Task.isCancelled || error is CancellationError {
                    listener.onGenerationDone()
                } else {
                    listener.onGenerationError(message: self.describe(error))
                }
            }
            self.generations[handle] = nil
        }
        generations[handle] = task
    }

    func cancelGeneration(handle: String) {
        if let conversation = conversations[handle] {
            try? conversation.cancel()
        }
        generations[handle]?.cancel()
        generations[handle] = nil
    }

    // MARK: - Helpers

    private func initialMessages(fromHistoryJson json: String) -> [Message] {
        guard
            let data = json.data(using: .utf8),
            let array = (try? JSONSerialization.jsonObject(with: data)) as? [[String: Any]]
        else { return [] }
        return array.compactMap { item in
            guard let role = item["role"] as? String, let text = item["text"] as? String else {
                return nil
            }
            return Message(text, role: role == "user" ? .user : .model)
        }
    }

    private func tools(for ids: [String]) -> [any Tool] {
        ids.compactMap { id in
            switch id {
            case "get_current_time": return GetCurrentTimeTool()
            case "calculate": return CalculateTool()
            case "set_theme": return SetThemeTool()
            case "set_language": return SetLanguageTool()
            case "random_number": return RandomNumberTool()
            case "device_info": return DeviceInfoTool()
            case "load_skill": return LoadSkillTool()
            default: return nil
            }
        }
    }

    private func describe(_ error: Error) -> String {
        let text = error.localizedDescription
        return text.isEmpty ? String(describing: error) : text
    }
}

File: lamityLlm/src/commonMain/kotlin/com/phamtunglam/lamity/llm/LlmBridge.kt
package com.phamtunglam.lamity.llm

/**
 * Contract between shared code and the platform LiteRT-LM runtime.
 *
 * Android implements it in this module's androidMain with the LiteRT-LM
 * Kotlin API ([AndroidLlmBridge]). iOS implements it in Swift
 * (iosApp/LlmBridge.swift) with the LiteRT-LM Swift API, so every method is
 * callback-based — Swift classes cannot implement Kotlin suspend members.
 */
interface NativeLlmBridge {
    /** Set by shared code; platform runtimes route model tool calls through it. */
    var toolExecutor: ToolExecutor?

    fun initializeEngine(setup: EngineSetup, callback: EngineCallback)
    fun closeEngine()

    fun createConversation(setup: ConversationSetup, callback: ConversationCallback)
    fun closeConversation(handle: String)

    fun sendMessage(handle: String, text: String, listener: GenerationListener)
    fun cancelGeneration(handle: String)
}

/** Executes a tool synchronously and returns the result JSON. */
interface ToolExecutor {
    fun executeTool(toolId: String, paramsJson: String): String
}

class EngineSetup(
    val modelPath: String,
    /** "cpu" or "gpu". */
    val backend: String,
    val maxTokens: Int,
    val cacheDir: String,
)

class ConversationSetup(
    val systemPrompt: String?,
    /** JSON array: [{"role":"user"|"model","text":"..."}] replayed when resuming. */
    val historyJson: String,
    /** Enabled tool ids; iOS maps these to its Swift tool structs. */
    val toolIds: List<String>,
    /** JSON array of {name,description,parameters} specs; Android builds OpenApiTools from it. */
    val toolSpecsJson: String,
    val topK: Int,
    val topP: Double,
    val temperature: Double,
)

interface EngineCallback {
    fun onEngineReady()
    fun onEngineError(message: String)
}

interface ConversationCallback {
    fun onConversationReady(handle: String)
    fun onConversationError(message: String)
}

interface GenerationListener {
    fun onChunk(text: String)
    /** Reasoning-channel delta (DeepSeek-R1 style); may never be called. */
    fun onThought(text: String)
    fun onGenerationDone()
    fun onGenerationError(message: String)
}
